File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool dfs(int i, vector<vector<int>>&adj, vector<int>&vis, vector<int>&path ){
        vis[i] = 1;
        path[i] = 1;
        for(auto it : adj[i]){
            if(!vis[it]) {
                if(dfs(it, adj, vis, path))
                return true;
            }
            else if(path[it]) return true;
        }
        path[i] = 0;
        return false;
    }
    bool canFinish(int numCourses, vector<vector<int>>& pre) {
       //number of tasks = numcourses
       //to complete a task B we need to complete A first , hence {B,A}, A-->B here

       //we will first create the adjacency list from the given prerequisites
       vector<vector<int>>adj(numCourses);
        for(int i = 0; i<pre.size(); i++){
            int a = pre[i][1];
            int b = pre[i][0];
            adj[a].push_back(b);
        }

        //now we will detect cycle in this graph (dfs)
        vector<int>vis(numCourses, 0);
        vector<int>path(numCourses,0);
        for(int i = 0 ; i<numCourses; i++){
        if(!vis[i]){
        if(dfs(i, adj, vis, path))
            return false;
    }
}

        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    # def dfs(self, i , adj, visited, path):
    #     visited[i] = 1
    #     path[i] = 1

    #     for neighbour in adj[i]:
    #         if not visited[neighbour]:
    #             if self.dfs(neighbour, adj, visited, path):
    #                 return True
    #         elif path[neighbour]: 
    #             return True
    #     path[i] = 0
    #     return False

    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # "using dfs"
        # 'creating adjancency list from given input'
        # adj = [[] for _ in range(numCourses)]
        # for a, b in prerequisites:
        #     adj[b].append(a)

        # 'detect a cycle in graph'
        # visited = [0] * numCourses
        # path = [0] * numCourses
        # for i in range(numCourses):
        #     if not visited[i]:
        #         if self.dfs(i, adj, visited, path):
        #             return False
                
        # return True

        #using bfs, kahns algorithm
         # Build adjacency list
        adj = [[] for _ in range(numCourses)]

        # Indegree of each course
        indegree = [0] * numCourses

        # [course, prerequisite]
        # prerequisite -> course
        for course, prerequisite in prerequisites:
            adj[prerequisite].append(course)
            indegree[course] += 1
        
        q = deque() #for 0 indegree
        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)
        count = 0
        #kahn's algorithm 
        while q :
            node = q.popleft()
            count += 1
            for neighbour in adj[node]:
                indegree[neighbour] -= 1

                if indegree[neighbour] == 0:
                    q.append(neighbour)

        return count == numCourses
        